es6-shim.js

```diff
@@ -2097,6 +2170,257 @@
+          }
+
+          return func.apply(this, arguments);
+        };
+      };
+
+      var internal_get = function get(target, key, receiver) {
+        var desc = Object.getOwnPropertyDescriptor(target, key);
+
+        if (!desc) {
+          var parent = Object.getPrototypeOf(target);
+
+          if (parent === null) {
+            return undefined;
+          }
+
+          return internal_get(parent, key, receiver);
+        }
+
+        if ('value' in desc) {
+          return desc.value;
+        }
+
+        if (desc.get) {
+          return desc.get.call(receiver);
+        }
+
+        return undefined;
+      };
+
+      var internal_set = function set(target, key, value, receiver) {
+        var desc = Object.getOwnPropertyDescriptor(target, key);
+
+        if (!desc) {
+          var parent = Object.getPrototypeOf(target);
+
+          if (parent !== null) {
+            return internal_set(parent, key, value, receiver);
+          }
+
+          desc = {
+            value: void 0,
+            writable: true,
+            enumerable: true,
+            configurable: true
+          };
+        }
+
+        if ('value' in desc) {
+          if (!desc.writable) {
+            return false;
+          }
+
+          if (!ES.TypeIsObject(receiver)) {
+            return false;
+          }
+
+          var existingDesc = Object.getOwnPropertyDescriptor(receiver, key);
+
+          if (existingDesc) {
+            return __defineOwnProperty(receiver, key, {
+              value: value
+            });
+          } else {
+            return __defineOwnProperty(receiver, key, {
+              value: value,
+              writable: true,
+              enumerable: true,
+              configurable: true
+            });
+          }
+        }
+
+        if (desc.set) {
+          desc.set.call(receiver, value);
+          return true;
+        }
+
+        return false;
+      };
+
+      // Some Reflect methods are basically the same as
+      // those on the Object global, except that a TypeError is thrown if
+      // target isn't an object. As well as returning a boolean indicating
+      // the success of the operation.
+      defineProperties(Reflect, {
+
+        // Apply method in a functional form.
+        apply: ES.Call,
+
+        // New operator in a functional form.
+        construct: function construct(constructor, args) {
+          if (!ES.IsCallable(constructor)) {
+            throw new TypeError('First argument must be callable.');
+          }
+
+          return ES.Construct(constructor, args);
+        },
+
+        defineProperty: wrapObjectFunction(Object.defineProperty),
+
+        // When deleting a non-existant or configurable property,
+        // true is returned.
+        // When attempting to delete a non-configurable property,
+        // it will return false.
+        deleteProperty: function deleteProperty(target, key) {
+          if (!ES.TypeIsObject(target)) {
```

Comment:
Could we wrap `function deleteProperty` in `throwUnlessTargetIsObject` instead, so it's consistent?